src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java

```diff
@@ -89,6 +95,69 @@ protected StructuredRecord computeNext() {
       }
     }
 
+    // get the next record that is available to compute the end time, it must be a record with same id
+    // that will not get dedup, or null if cur record is last record/
+    // This is needed because of the following scenarios:
+    // 1(cur): id: 1, name: john, isTarget: true (should be preserved)
+    // 2: id: 1, name: john, isTarget: false (should get dedup)
+    // 3: id: 1, name: sam, isTarget: false (should get dedup)
+    // 4: id: 1, name: sam, isTarget: false (should be preserved)
+    // here first and fourth record will be preserved, so when we compute the end date of 1st record, we should use
```

Comment:
I'm not really following the logic here. Why do we need nextNext and why do we need 2 dedup loops? Why isn't the change in the loop above enough? Shouldn't cur be set to record 1 and prev be set to record 4 at that point?

Reply:
No, prev should always be before the cur record. 
The first loop is to move the iterator to find the correct cur, the second loop is to move the iterator to find the correct next.

If you consider the the following example:
1: id: 1, name: john, isTarget: false (should get dedup)
2: id: 1, name: john, isTarget: true(should be preserved)
3: id: 1, name: john, isTarget: false (should get dedup)
4: id: 1, name: sam, isTarget: false (should get dedup)
5: id: 1, name: sam, isTarget: false (should be preserved)

After the first loop, #2 will be found as cur record, at this time, prev points to #1, and next points to #3. But #3 is a record should get discarded because it is same as #2, which is prev record. So a second loop is needed to check a next record which either a record which is in target table, or a record that should not be discarded, which means it is must be different from the cur record(#2), but also different from the following records.(#5) Thats why we need to have nextNext and a second loop. 
I tried to put the code in one loop, and the code will be super hard to follow, since that loop involves updating prev, cur, next, nextNext at the same time as well as comparing the prev and nextNext. That's why I break down this to a second loop to make logic easier to follow

Reply:
@albertshau I updated the doc as well as updating this comment about the behavior, we can discuss about it if you want.

Reply:
I'm still having a hard time following this logic. It seems difficult to understand without actually stepping through examples. I think it would be cleaner if we did a few things:

1.  Wrap the Iterator given in the constructor with a PeekingIterator (from guava). Then you can just peek() instead of maintaining extra state yourself.

2. Refactor the dedupe check into a separate method

3. Use descriptive variable names instead of calling ._1() and ._2() inline.